### PR TITLE
add medallia styling as separate output

### PR DESCRIPTION
**Pull request by rianfowler**
## GitHub Issue

[Host Medallia CSS #13287](https://github.com/department-of-veterans-affairs/va.gov-team/issues/13287) 

## Description

Remove hash from Medallia CSS file in staging environment

## Testing done

1. Run `yarn build:webpack:prod`
   - Build time with sourcemaps = ~50 minutes 😱 
   - Build time without sourcemaps = ~15 minutes 😐 
1. Run `ls -l build/vagovprod/generated | grep medal` to find the medallia CSS file
1. Confirm there is a hash in the filename (for example, `va-medallia-styles.a9a525ad4ac2474ed8b4-1601477867140.css`)
1. Run `yarn build:webpack:staging`
1. Run `ls -l build/vagovstaging/generated | grep medal` to find the medallia CSS file
1. Confirm there is no hash in the filename (for example, `va-medallia-styles.css`)

## Acceptance criteria
- [ ]

## Definition of done
- [ ] Documentation has been updated, if applicable
- [x] A link has been provided to the originating GitHub issue (or connected to it via ZenHub)

### CODE REVIEW

File: config/webpack.config.js [hunk 1]
```diff
@@ -236,9 +241,14 @@ module.exports = env => {
       }),
 
       new MiniCssExtractPlugin({
-        filename: !isOptimizedBuild
-          ? '[name].css'
-          : `[name].[contenthash]-${timestamp}.css`,
+        moduleFilename: ({ name, entryModule }) => {
```
Comment (billfienberg) [inline config/webpack.config.js:244]: `moduleFilename` docs: https://webpack.js.org/plugins/mini-css-extract-plugin/#module-filename-option

Comment (billfienberg) [reply config/webpack.config.js:244]: Here are the keys that are on the object argument for the `moduleFilename` function:
```
[ 'id',
  'ids',
  'debugId',
  'name',
  'preventIntegration',
  'entryModule',
  '_modules',
  'filenameTemplate',
  '_groups',
  'files',
  'rendered',
  'hash',
  'contentHash',
  'renderedHash',
  'chunkReason',
  'extraAsync',
  'removedModules' 
]
```

Comment (billfienberg) [reply config/webpack.config.js:244]: https://webpack.js.org/configuration/output/#template-strings

image

**Comment by billfienberg** [comment]
Copying a Slack conversation here for visibility
>**Rian**: tl;dr we need to push just this css file out without a hash on the filename. I have figured out how to configure webpack to do that but the output is not matching the master branch output.
>
>once we can publish the file to staging, we need to give it to insights team to add to the medallia iframe settings. it will load the css dynamically and we may need to do some tweaks to make the survey modal look like the design.
>
>**Bill**: Which output? The entire `build` folder? The `generated` folder? The medalia specific file?
>
>**Rian**: the generated folder
>sorry i should have been more specific about this
>i’ve been running this command: `NODE_ENV=production yarn build:webpack --env.buildtype vagovprod`
>on my branch vs master branch and comparing the output of build/vagovprod
>that process takes forever which is the hard part 
>that argument for the `moduleFileName` is not documented anywhere- there are 3 different hashes with different names in that available as an argument and i may not be using the right one
>
>**Bill**: I suspected it was `generated`, but I've been diffing the entire build to be safe. And those builds are slow.
>
>**Rian**: what i saw was that my branch build output more files than the master branch- more things without hashes which was confusing
>yeah and you have to make sure your master / local branches are on the same commit
>i would expect the hashes to be the same but i may not be using the same hash `renderedHash` may not be what `filename` is using
>so you might want to console log out that object to see what all of the options on i should have captured it somewhere

File: package.json [hunk 2]
```diff
@@ -11,6 +11,7 @@
     "build:content": "node --max-old-space-size=8192 script/build-content.js",
     "build:content:test": "node script/test-cms-export.js",
     "build:webpack": "NODE_OPTIONS=--max-old-space-size=4096 webpack --config config/webpack.config.js",
+    "build:webpack:prod": "NODE_ENV=production yarn build:webpack --env.buildtype vagovprod",
```
Comment (billfienberg) [inline package.json:14]: This is intended to be temporary. I just added it so I don't mistakenly run `yarn build` and end up with a `build/localhost` folder.

**Comment by billfienberg** [comment]
Disabling source maps reduced my build from ~3,000 seconds (~50 minutes 😱 ) to 625.29s. (~10.5 minutes 😐 ).

File: config/webpack.config.js [hunk 3]
```diff
@@ -13,14 +13,16 @@ const ManifestPlugin = require('webpack-manifest-plugin');
 const WebpackBar = require('webpackbar');
 
 const headerFooterData = require('../src/platform/landing-pages/header-footer-data.json');
-const BUCKETS = require('../src/site/constants/buckets');
+// const BUCKETS = require('../src/site/constants/buckets');
```
Comment (billfienberg) [inline config/webpack.config.js:16]: Temporary - will be deleted before merging

File: config/webpack.config.js [hunk 4]
```diff
@@ -148,7 +153,7 @@ module.exports = env => {
             },
             {
               loader: 'sass-loader',
-              options: { sourceMap: true },
+              // options: { sourceMap: true },
```
Comment (billfienberg) [inline config/webpack.config.js:156]: Temporary - will be deleted before merging

File: config/webpack.config.js [hunk 5]
```diff
@@ -214,7 +219,7 @@ module.exports = env => {
           },
           // cache: true,
           parallel: 3,
-          sourceMap: true,
+          // sourceMap: true,
```
Comment (billfienberg) [inline config/webpack.config.js:222]: Temporary - will be deleted before merging

File: config/webpack.config.js [hunk 6]
```diff
@@ -373,14 +383,14 @@ module.exports = env => {
...
   if (isOptimizedBuild) {
-    const bucket = BUCKETS[buildOptions.buildtype];
-
-    baseConfig.plugins.push(
-      new webpack.SourceMapDevToolPlugin({
-        append: `\n//# sourceMappingURL=${bucket}/generated/[url]`,
-        filename: '[file].map',
-      }),
-    );
+    // const bucket = BUCKETS[buildOptions.buildtype];
```
Comment (billfienberg) [inline config/webpack.config.js:386]: Temporary - will be deleted before merging

File: config/webpack.config.js [hunk 7]
```diff
@@ -373,14 +383,14 @@ module.exports = env => {
...
+
+    // baseConfig.plugins.push(
```
Comment (billfienberg) [inline config/webpack.config.js:388]: Temporary - will be deleted before merging

**Comment by billfienberg** [comment]
I wonder if we should be using `contentHash` instead of `renderedHash`. 

>TLDR: Use [contenthash] instead of [chunkhash] to ensure that every asset (output file) has a stable hash based on it's contents (only). - https://github.com/webpack/webpack/issues/7179#issuecomment-386132702

**Comment by billfienberg** [comment]
I wonder if we should try adding `optimization: moduleIds: 'hashed'`:

>Now, despite any new local dependencies, our vendor hash should stay consistent between builds: - https://webpack.js.org/guides/caching/#module-identifiers

>`moduleIds: 'deterministic'` was added in webpack 5 and `moduleIds: 'hashed'` is deprecated in favor of it. - https://webpack.js.org/configuration/optimization/#optimizationmoduleids

File: package.json (hunk 2 above)
Comment (rianfowler) [reply package.json:14]: looks good

**Comment by billfienberg** [comment]
I'm seeing lots of `*.undefined-${timestamp}.css` when using `${contenthash}`. 🤔 

image

**Comment by billfienberg** [comment]
🤦‍♂️ 

The MiniCssExtractPlugin plugin options seem to be documented in a JSON file. 

https://github.com/webpack-contrib/mini-css-extract-plugin/blob/master/src/plugin-options.json

File: package.json [hunk 8]
```diff
@@ -11,6 +11,8 @@
...
+    "debug:webpack:prod": "NODE_ENV=production NODE_OPTIONS=--max-old-space-size=4096 node --inspect --inspect-brk ./node_modules/webpack/bin/webpack.js --config config/webpack.config.js --env.buildtype vagovprod",
```
Comment (billfienberg) [inline package.json:15]: 1. Run `debug:webpack:prod`
1. Open chrome://inspect
1. Click **Inspect** for the remote target (./node_modules/webpack/bin/webpack.js)

debug-node

Reference: https://webpack.js.org/contribute/debugging/

File: config/webpack.config.js [hunk 9]
```diff
@@ -236,9 +241,14 @@ module.exports = env => {
       }),
 
       new MiniCssExtractPlugin({
-        filename: !isOptimizedBuild
-          ? '[name].css'
-          : `[name].[contenthash]-${timestamp}.css`,
+        moduleFilename: chunk => {
```
Comment (billfienberg) [inline config/webpack.config.js:244]: https://github.com/webpack-contrib/mini-css-extract-plugin/issues/67#issuecomment-520795293

**Comment by billfienberg** [comment]
Related tickets:

- [Upgrade to Webpack 5](https://github.com/department-of-veterans-affairs/va.gov-team/issues/11510) 
- [Modularize the Webpack config](https://github.com/department-of-veterans-affairs/va.gov-team/issues/2796)

**Comment by billfienberg** [comment]
`build/vagovprod` = `va-medallia-styles.a9a525ad4ac2474ed8b4-1601477867140.css`
`build/vagovstaging` = `va-medallia-styles.css`

image

File: config/webpack.config.js [hunk 10]
```diff
@@ -236,9 +241,20 @@ module.exports = env => {
...
+          const isMedalliaStyleFile = name === vaMedaliaStylesFilename;
+          const isStandardFile = !isMedalliaStyleFile;
+          const showHash = isOptimizedBuild && isStandardFile;
+
+          if (showHash) return standardFilename;
+
+          const isStaging =
+            buildOptions.buildtype === ENVIRONMENTS.VAGOVSTAGING;
+
+          return isStaging ? `${name}.css` : standardFilename;
```
Comment (U-DON) [inline config/webpack.config.js:256]: Is this for testing on staging first?

File: config/webpack.config.js [hunk 11]
```diff
@@ -21,6 +21,8 @@ const {
   getWebpackEntryPoints,
 } = require('./manifest-helpers');
 
+const vaMedaliaStylesFilename = 'va-medallia-styles';
```
Comment (U-DON) [inline config/webpack.config.js:24]:
```suggestion
const vaMedalliaStylesFilename = 'va-medallia-styles';
```

Comment (billfienberg) [reply config/webpack.config.js]: 🤦‍♂️  I made that change locally earlier today, but forgot to push it. 

0207819b6edddcb09f8f52f2b5f2902d7f9c15e1

File: config/webpack.config.js (hunk 10 above)
Comment (billfienberg) [reply config/webpack.config.js:256]: That's correct. 

>once we can publish the file to staging, we need to give it to insights team to add to the medallia iframe settings. - https://github.com/department-of-veterans-affairs/vets-website/pull/14312#issuecomment-700067751

**Review by U-DON — APPROVED** [review]

File: config/webpack.config.js [hunk 12]
```diff
@@ -21,6 +21,8 @@ const {
   getWebpackEntryPoints,
 } = require('./manifest-helpers');
 
+const vaMedalliaStylesFilename = 'va-medallia-styles';
```
Comment (rianfowler) [inline config/webpack.config.js:24]:
```suggestion
// TODO: refactor the other approach for creating files without the hash so that we're only doing that in the webpack config: https://github.com/department-of-veterans-affairs/vets-website/blob/a012bad17e5bf024b0ea7326a72ae6a737e349ec/src/site/stages/build/plugins/process-entry-names.js#L35
const vaMedalliaStylesFilename = 'va-medallia-styles';
```

File: config/webpack.config.js [hunk 13]
```diff
@@ -236,9 +241,20 @@ module.exports = env => {
...
+        moduleFilename: chunk => {
+          const { name } = chunk;
+          const standardFilename = `${name}.[contenthash]-${timestamp}.css`;
+          const isMedalliaStyleFile = name === vaMedaliaStylesFilename;
+          const isStandardFile = !isMedalliaStyleFile;
+          const showHash = isOptimizedBuild && isStandardFile;
+
+          if (showHash) return standardFilename;
+
+          const isStaging =
```
[the remark this answers is not in the document]
Comment (billfienberg) [reply config/webpack.config.js]: replaced temp variable with environment helper in 0412326a8c5d81c670da44f33c43c7261e83ff1a

File: config/webpack.config.js (hunk 10 above)
Comment (billfienberg) [reply config/webpack.config.js]: >I think we should have the staging check before any return statements. otherwise this logic is fine.

@rianfowler, I don't have an objection to moving the staging check before any `return`s. That being said, is the Medallia CSS file the only file that cares about the staging check? If so, what is the value of performing the staging check for every file?

Comment (billfienberg) [reply config/webpack.config.js]: >Now that I'm getting a second look at this, it seems like dev would keep the hash in the name (not optimized build, so showHash is falsy). Not sure how much this matters.

@U-DON, my intent for the code was:
1. Determine if it's the Medallia CSS file.
1. If not, return our normal filename format.
1. If so, determine if it's the staging* environment.
   * My understanding of the requirements is that I should be checking only for staging. If I should be checking for additional non-prod environments, then I need to update my code. 
1. If staging, exclude the hash. 
1. If non-staging, include the hash.

File: config/webpack.config.js [hunk 14]
```diff
@@ -236,9 +243,17 @@ module.exports = env => {
...
+        moduleFilename: chunk => {
+          const { name } = chunk;
+          const standardFilename = `${name}.[contenthash]-${timestamp}.css`;
+          const isMedalliaStyleFile = name === vaMedalliaStylesFilename;
+          const isStandardFile = !isMedalliaStyleFile;
+          const showHash = isOptimizedBuild && isStandardFile;
+
+          if (showHash) return standardFilename;
+
+          return environment.isStaging() ? `${name}.css` : standardFilename;
```
Comment (U-DON) [inline config/webpack.config.js:255]: >> Now that I'm getting a second look at this, it seems like dev would keep the hash in the name (not optimized build, so showHash is falsy). Not sure how much this matters.
> 
> @U-DON, my intent for the code was:
> 
> 1. Determine if it's the Medallia CSS file.
> 1. If not, return our normal filename format.
> 1. If so, determine if it's the staging* environment.
>     - My understanding of the requirements is that I should be checking only for staging. If I should be checking for additional non-prod environments, then I need to update my code.
> 1. If staging, exclude the hash.
> 1. If non-staging, include the hash.

Restarting thread on new changes so the discussion doesn't get lost from an outdated snippet.

If the requirements is only for staging, that's fair. I just wanted to point out that this looks like it woud add hashes for dev, which is a deviation from the original code.

If we want to be explicit about "hash only for Medallia, only in staging", then would it be clearer to write it like this?

```suggestion
          const standardFilename = `${name}.[contenthash]-${timestamp}.css`; 
          const isMedalliaStyleFile = name === vaMedalliaStylesFilename; 
          
          if (isMedalliaStyleFile && environment.isStaging()) return `${name}.css`;
          
          return isOptimizedBuild ? standardFilename : `${name.css}`;
```

Comment (billfienberg) [reply config/webpack.config.js:255]: This looks a lot cleaner. Thanks for the suggestion.

File: config/webpack.config.js (hunk 13 above)
[the remark this answers is not in the document]
Comment (billfienberg) [reply config/webpack.config.js:249]: I went back to using a temp variable instead of the environment helper function in f832dd3f7bd09395c86b9ea15a1b81f555e7cdb2. 

The [build was failing](https://app.circleci.com/pipelines/github/department-of-veterans-affairs/vets-website/14370/workflows/7637a22e-c24d-480d-b719-c20fdf082156/jobs/110434/parallel-runs/0/steps/0-105). I suspect it's because the helper function uses an `import` statement, and we're using `node@10`, which doesn't support `import` statements. 
```
import ENVIRONMENTS from 'site/constants/environments';
       ^^^^^^^^^^^^
SyntaxError: Unexpected identifier
```

image

**Review by U-DON — APPROVED** [review]
